File: 117_jump-game-ii/jump-game-ii.java
/*
@Copyright:LintCode
@Author:   hanqiao
@Problem:  http://www.lintcode.com/problem/jump-game-ii
@Language: Java
@Datetime: 16-05-29 07:46
*/

public class Solution {
    /**
     * @param A: A list of lists of integers
     * @return: An integer
     */
    public int jump(int[] A) {
        // write your code here
        if (A == null || A.length == 0) {
            return 0;
        }
        //dynamic programming
        int[] minSteps = new int[A.length];
        minSteps[0] = 0;
        for (int i = 1; i < A.length; i++) {
            minSteps[i] = minSteps[i - 1] + 1;
            for (int j = 0; j < i; j++) {
                if (A[j] + j >= i) {
                    minSteps[i] = minSteps[j] + 1;
                    break;
                }
            }
        }
        return minSteps[A.length - 1];
    }
}

File: 117_jump-game-ii/jump-game-ii.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   hanqiao
@Problem:  http://www.lintcode.com/problem/jump-game-ii
@Language: Python
@Datetime: 16-05-29 08:06
'''

class Solution:
    # @param A, a list of integers
    # @return an integer
    def jump(self, A):
        # write your code here
        if A is None or len(A) == 0:
            return 0
        start = 0
        end = 0
        step = 0
        #greedy algorithm
        # O(1) space and O(n^2) worst case time
        while end < len(A) - 1:
            step = step + 1
            farthest = end
            for i in range(start, end + 1):
                if A[i] + i > farthest:
                    farthest = A[i] + i
            start = end + 1
            end = farthest
        return step
            